### PR TITLE
[JBPM-9436] AMQ Streams (Kafka) listener functionality / consumer

### PR DESCRIPTION
**Pull request by fjtirado**
Adding method getSignalsMetadata to ProcessDefinition interface

Merge with [#droolsjbpm-integration:2288](https://github.com/kiegroup/droolsjbpm-integration/pull/2288)

**JIRA**: 

[JBPM-9436](https://issues.redhat.com/browse/JBPM-9436)

### CODE REVIEW

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
jenkins retest this

**Review by afalhambra — COMMENTED** [review]
Only minor comments, but it looks pretty good

File: jbpm-event-emitters/jbpm-event-emitters-kafka/pom.xml
Comment (fjtirado) [reply jbpm-event-emitters/jbpm-event-emitters-kafka/pom.xml:32]: This was discussed with @elguardan, but yes, because this is intended to be used only in kie-server, that now will always have the dependency

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java:253]: Probably, but since Im not sure and they are not related with this JIRA, I refrained myself and decided to not clear them

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/model/ProcessDefinition.java
Comment (fjtirado) [reply jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/model/ProcessDefinition.java:95]: not the same, id and name might be differ and this method is returning the ids

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/SignalDescImpl.java
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/SignalDescImpl.java:32]: signal is not expected to be null there, if it is a runtime will be thrown anyway

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/SignalDescImpl.java
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/SignalDescImpl.java:36]: same rationale, message must not be null, since this is unexpected, it is acceptable to have the runtime

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/SignalDescImplTest.java
Comment (fjtirado) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/SignalDescImplTest.java:34]: ok, I technically "abuse" pepe here, but because I want to emphasize that everything is equals and despite of it, since one SignalDesc come from Signal object and another SignalDesc come from Message, theere are two instances at the end in the set. Changed anyway

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java
Comment (fjtirado) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java:603]: Our engine is supporting undefined signals (meaning that you can put a signal name in an intermediate event without a signal definition). I do not want to enforce that (because process that are now considered validated, wont be anymore), so I have to add this comparison in code

**Review by afalhambra — APPROVED** [review]
Thanks @fjtirado - very nice work

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java:603]: Any time you create a signal in the bpmn2 modeler should create a signal element. So this should not be needed.  
This is actually hiding a problem in the bpmn2 modeler IMO if that is happening.

There are few cases where signals are used to implement something else. They should not be defined as signals but messages.
Start Message (implemented as signal)
Boundary events (messages BoundaryEventHandler)
catching events (messages)

this should be removed.



In spite they are implemented as signal you need to make a difference in objects there.

File: jbpm-event-emitters/jbpm-event-emitters-kafka/pom.xml
Comment (elguardian) [reply jbpm-event-emitters/jbpm-event-emitters-kafka/pom.xml:32]: emitters can be used in spring boot envioronment. plz remove this.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java:68]: this is redundant as we have the signal object here already.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java:70]: This objects are coming from bpm2 parser. They should not be here.
you have an object already for this SignalDesc.
I dont see the counter part MessageDesc

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/bpmn2/ProcessDescriptor.java:185]: it is not signal metadata... it is the signal element it self. change for a more descriptive name. i.e signallement or signalobject.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/ProcessAssetDesc.java
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/ProcessAssetDesc.java:219]: same as before. change to a more descriptive name. this is not the metadata but  the element itself.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/ProcessAssetDesc.java
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/ProcessAssetDesc.java:223]: this should not be here. create a dependency with the parser.
try to put this logic in ServicesProcessDataEventListener

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/model/SignalDesc.java
Comment (elguardian) [inline jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/model/SignalDesc.java:28]: there is no signal type only id, name and structure ref.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/ProcessAssetDesc.java
Comment (elguardian) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/model/ProcessAssetDesc.java:231]: this should not be here. create a dependency with the parser.
try to put this logic in ServicesProcessDataEventListener

File: jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/model/SignalType.java
Comment (elguardian) [inline jbpm-services/jbpm-services-api/src/main/java/org/jbpm/services/api/model/SignalType.java:1]: not required.

**Review by elguardian — CHANGES_REQUESTED** [review]
some changes required.

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java
Comment (fjtirado) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java:603]: There are some processes with have a signal ref, but not define signal object, this code is trying to completate that scenario
https://github.com/kiegroup/jbpm/blob/master/jbpm-services/jbpm-kie-services/src/test/resources/repo/processes/general/signal.bpmn

File: jbpm-event-emitters/jbpm-event-emitters-kafka/pom.xml
Comment (elguardian) [reply jbpm-event-emitters/jbpm-event-emitters-kafka/pom.xml:32]: as per  @fjtirado this is following the same approach as ES.

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java
Comment (fjtirado) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/AbstractNodeHandler.java:603]: After talking with Enrique, this non defined signals should not be exposed externally

**Review by fjtirado — COMMENTED** [review]
getMessageDesc added. Initial implementation intention was to  take benefit of the fact that currently MEssage and Signal hold the same data, but as Enrique pointed out, this will prevent further divergence between them.

**Comment by fjtirado** [comment]
@kiegroup/gatekeepers please merge this one

**Comment by mareknovotny** [comment]
@fjtirado :o, please if you have cross repository PR set be sure all is ready to merge, I can't see the PR on droolsjbpm-integration repository review and approved.
Next in case of multirepo PRs there should be also triggered full downstream build
Without FDB  and all tests ran there we cannot merge it as it can cause the followup problems in other repositories.

jenkins do fdb

**Comment by elguardian** [comment]
@fjtirado this requires approval before asking for merge.

**Review by elguardian — APPROVED** [review]
LGTM.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/MessageDescImplTest.java
Comment (afalhambra) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/MessageDescImplTest.java:1]: Shouldn't this test class be located under a package `org.jbpm.kie.services.impl.model` just to follow same package structure for the class to test?

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/SignalDescImplTest.java
Comment (afalhambra) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/SignalDescImplTest.java:1]: Shouldn't this test class be located under a package `org.jbpm.kie.services.impl.model` just to follow same package structure for the class to test?

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java
Comment (afalhambra) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java:534]: Checked exception is not thrown within this method
```suggestion
    public void testSignalsDesc() {
```

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java
Comment (afalhambra) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java:540]: Should be changed to:
```suggestion
        assertFalse("Signal description should not be empty", signals.isEmpty());
```

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java
Comment (afalhambra) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java:544]: I would add an assert for `structRef` as well, even if it is null, just to cover all possible attributes
```suggestion
        assertEquals("MySignal", signal.getName());
        assertNull("Structure Ref is expected to be null", signal.getStructureRef());
```

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java
Comment (afalhambra) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/BPMN2DataServicesReferencesTest.java:554]:
```suggestion
        assertFalse("Message description should not be empty", signals.isEmpty());
```

**Review by afalhambra — CHANGES_REQUESTED** [review]
Minor comments, specially in the test cases, but nothing major.

It looks very good to me otherwise.

**Review by afalhambra — APPROVED** [review]

**Pull request merged**
